### DISCUSSION TITLE
Real Time Machine Learning with AnyLogic Model

### DISCUSSION THREAD

**Discussion by swaneetrao** [the-anylogic-company/AnyLogic-Pypeline#29] [Q&A] [answered]
Hello Tyler,

I am working on a resource allocation business AnyLogic model. I want to extract data during runtime of simulation, process the data using some ML models and feed the ML output data back to my simulation and vary parameters dynamically in the real time, so that my simulation model will act as feedback and environment for my ML model ( Example : like increase/decrease the population of my transporter fleet, increase the speed of transporters etc., depending on my ML model's output to increase efficiency of my AnyLogic model in real time )  and implement an active ecosystem of AnyLogic model with real time machine learning. Is this possible using Pypeline (pycommunicator) ? If yes, please help me out with the steps on how I can implement those. As I am using AnyLogic for the first time I am bit confused, it will be of great help if you guide me in the right direction.

Thanks,
Swaneet

**Comment by t-wolfeadam** [comment, marked as answer]
Hi, sorry about the delay -- Pypeline is just a mechanism to access Python from Java (i.e., AnyLogic) so there's nothing stopping you from using it in this way. Though, the complicated parts your request concerns the AL and ML models themselves; the connection itself is rather straight forward and is what I can help best with once you have the two ready.

I suggest focusing on building each part separately, for the purposes of being able to efficiently debug it (e.g., restarting your Python code is much faster than restarting the AL model that calls the Python code).

- You can build the core of the AL model first without the dynamic interactions that the ML model will later control (as you can add those later as their own events). 
- Then you can focus on setting up the Python code and tuning the ML model's parameters. With this, you can generate some data for sampling purposes; exporting this can be done using AL's Excel or Text objects or Pypeline for its JSON capabilities. 
- Lastly would be the connection between the two, which is just some simple function calls (see the Wiki for examples).

As a final remark, I get that you want to call an ML model from AnyLogic. Though, if it's the case you're not wanting to retrain it in real-time, I'll mention there's also a ONNX wrapper library I have, specifically for calling trained ML models, which will be much more time-efficient than Pypeline.

**Reply by swaneetrao** [reply]
Hello , Thanks for your prompt reply. I understand your point.
There have been some changes in the constraints and I need to use reinforcement learning instead of traditional model based machine learning. Can you please comment on that on how to implement RL in AnyLogic Model.
Suggestions will much appreciated.

Thanks & Regards,
Swaneet

**Reply by t-wolfeadam** [reply]
I highly suggest using Microsoft's Project Bonsai for scalable and robust training. Some references on that:
* Official documentation: https://docs.microsoft.com/en-us/bonsai/product/
* AnyLogic/Bonsai workflow documentation: https://microsoft.github.io/bonsai-anylogic/index.html

However if you're looking for a Python implementation on a local machine (simpler use cases), there's the Alpyne library I developed. The caveat to using currently is that it's in a semi-stable beta. You can find more info at: https://github.com/t-wolfeadam/Alpyne

**Reply by swaneetrao** [reply]
Thanks a lot for your suggestion. I will go through the same and try to implement using above mentioned platforms. I will get back to you incase of any queries.
